test/README.md

```diff
@@ -0,0 +1,125 @@
+# Testing Guidelines #
+
+## Rationale ##
+
+Testing is a necessary aspect of any software development process. There are
+many ways to describe the quality of a test suite (both quantitative and
+qualitative). In addition to these properties, a high quality test suite still
+necessitates clear and coherent organisation.
+
+Whilst the test suite for Perpetual Pools exhibits excellent coverage metrics
+and captures the intended behaviour of the system rather well, it is currently
+organised very poorly (inconsistent naming, inconsistent file structure,
+inconsistent use of testing directives, etc.).
+
+This document aims to clarify how we conduct testing in the Perpetual Pools
+system and act as a procedural guide for how to write additional tests when
+contributing to the codebase. Specifically, this document aims to address the
+following questions:
+
+ - How should tests be laid out? File-per-contract, etc.
+ - What testing frameworks do we use?
+ - How should the tests themselves look?
+
+## Test Suite Structure ##
+
+### All Tests Under `test` ###
+
+All tests -- regardless of their type (e.g., unit, integration, etc.) -- must
+reside under the `test` directory.
+
+### One-File-One-Contract ###
+
+All unit tests should be arranged such that all tests for a given module are in
+the same file. The most obvious choice for what constitutes a "module" is a
+smart contract.
+
+For example, if the (fictional) contract `PermissionedDeployer` is being added
+to the codebase, its unit tests should reside in a file called
+`PermissionedDeployer.spec.ts` under the `test` directory.
+
+## Test Layout ##
+
+### Use BDD Testing Nomenclature ###
+
+Perpetual Pools uses [Mocha](https://mochajs.org) as its testing framework.
+Broadly speaking, Mocha exposes three different primitives for writing tests:
+
+ - `describe`
+    - Used to *describe* a given component of the system being tested
+    - Can be nested
+ - `context`
+    - Used to describe the situation being considered
+ - `it`
+    - Used to make an assertion about the behaviour of the system
+
+Additionally, the ability to nest these primitives must be used eagerly. For
+example,
+
+```typescript
+describe("PermissionedDeployer", async () => {
+    describe("get", async () => {
```

Comment:
Why would this be `describe` and not `context`? Base on your description above,  `describe` should describe a given component, but I would say this is not describing anything, but instead giving context as to what is being tested.

On that same note, wouldn't the top level `describe("PermissionedDeployer",...)` be more suited for `context`, given that it is basically saying "The context of this test is the PermissionedDeployer contract", rather than describing anything. You could argue that it is `describe` because it is describing what will be tested here but I don't know if that is compatible with the above description.

Reply:
The [relevant documentation for Mocha](https://mochajs.org/#bdd) seems to suggest this nesting is the intended ordering. At the same time, it says that `context` is just an alias to `describe` but I think this is more of an internal design choice of Mocha's implementation rather than a BDD thing.